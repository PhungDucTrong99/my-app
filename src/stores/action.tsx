import { v4 as uuidv4 } from "uuid";

export const addUsers = (data: any) => {
  return {
    type: "user/addUsers",
    payload: data,
  };
};

export const _saveQuestion = (data: any) => {
  return (dispatch: any) => {
    return new Promise<string>((resolve, reject) => {
      const id = uuidv4().toString();
      dispatch({
        type: "user/addQuestion",
        payload: {
          ...data,
          id: id,
        },
      });
      resolve(id);
    });
  };
};

export const _saveQuestionAnswer = (data: any) => {
  return (dispatch: any) => {
    return new Promise<string | void>((resolve, reject) => {
      dispatch({
        type: "user/handleAnswer",
        payload: data,
      });
      resolve();
    });
  };
};

export const isLogin = (id: any) => {
  return {
    type: "user/isLogin",
    payload: id,
  };
};

export const isLogout = (id: any) => {
  return {
    type: "user/isLogout",
    payload: id,
  };
};

export const Login = (payload: any) => {
  return (dispatch: any) => {
    return new Promise<string | void>((resolve, reject) => {
      dispatch({
        type: "user/Login",
        payload: { username: payload.username, password: payload.password },
      });
      resolve();
    });
  };
};
