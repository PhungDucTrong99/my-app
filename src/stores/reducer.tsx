// import {Image1, Image2, Image3 } from "../img";
import { AnyAction } from "redux";
import React from "react";
import { Image1, Image2, Image3 } from "../img/until";

interface User {
  id: string;
  username: string;
  image: JSX.Element;
  password: string;
  role: string;
  status: string;
  totalPools: number;
  answers: string[];
  question: string[];
}

interface Question {
  id: string;
  author: string;
  createtime: string;
  optionOne: {
    votes: string[];
    text: string;
  };
  optionTwo: {
    votes: string[];
    text: string;
  };
}

export interface State {
  isLoading: {
    status: boolean;
  };
  user: { [key: string]: User };
  listQuestion: { [key: string]: Question };
}

const initialState: State = {
  isLoading: {
    status: false,
  },
  user: {
    Cat: {
      id: "000",
      username: "Cat",
      image: <Image1 />,
      password: "070699",
      role: "admin",
      status: "inActive",
      totalPools: 4,
      answers: ["become a superhero"],
      question: ["6ni6ok3ym7mf1p33lnez"],
    },
    Fox: {
      id: "111",
      username: "Fox",
      image: <Image2 />,
      password: "070699",
      role: "admin",
      status: "inActive",
      totalPools: 1,
      answers: ["become a superhero"],
      question: ["6ni6ok3ym7mf1p33lnez"],
    },
    Gorilla: {
      id: "222",
      username: "Gorilla",
      role: "admin",
      image: <Image3 />,
      password: "070699",
      status: "inActive",
      totalPools: 1,
      answers: ["become a superhero"],
      question: ["6ni6ok3ym7mf1p33lnez"],
    },
  },
  listQuestion: {
    "8xf0y6ziyjabvozdd253nd": {
      id: "8xf0y6ziyjabvozdd253nd",
      author: "000",
      createtime: "11:03 AM : 13/12/2013",
      optionOne: {
        votes: [""],
        text: "have horrible short term memory",
      },
      optionTwo: {
        votes: [""],
        text: "have horrible long term memory",
      },
    },
    "6ni6ok3ym7mf1p33lnez": {
      id: "6ni6ok3ym7mf1p33lnez",
      author: "111",
      createtime: "11:03 AM : 12/12/2023",
      optionOne: {
        votes: [""],
        text: "become a superhero",
      },
      optionTwo: {
        votes: [""],
        text: "become a supervillian",
      },
    },
    am8ehyc8byjqgar0jgpub9: {
      id: "am8ehyc8byjqgar0jgpub9",
      author: "222",
      createtime: "11:03 AM : 12/12/2022",
      optionOne: {
        votes: [""],
        text: "be telekinetic",
      },
      optionTwo: {
        votes: [""],
        text: "be telepathic",
      },
    },
    loxhs1bqm25b708cmbf3g: {
      id: "loxhs1bqm25b708cmbf3g",
      author: "111",
      createtime: "11:03 AM : 12/12/2011",
      optionOne: {
        votes: [""],
        text: "be a front-end developer",
      },
      optionTwo: {
        votes: [""],
        text: "be a back-end developer",
      },
    },
  },
};

const reducer = (state = initialState, action: AnyAction): State => {
  switch (action.type) {
    case "user/isLogin":
      const updatedUser = Object.values(state.user).find(
        (user) => user.id === action.payload
      );

      if (updatedUser) {
        updatedUser.status = "Active";
      }

      return {
        ...state,
        user: { ...state.user },
      };

    case "user/Login":
      const loginUser = Object.values(state.user).find(
        (user) =>
          user.password === action.payload.password &&
          user.username === action.payload.username
      );

      if (loginUser) {
        loginUser.status = "Active";
      }

      return {
        ...state,
        user: { ...state.user },
      };

    case "user/isLogout":
      const updatedUserLogout = Object.values(state.user).find(
        (user) => user.id === action.payload
      );

      if (updatedUserLogout) {
        updatedUserLogout.status = "inActive";
      }

      return {
        ...state,
        user: { ...state.user },
      };

    case "user/addQuestion":
      const id = action.payload.id;

      return {
        ...state,
        listQuestion: {
          ...state.listQuestion,
          [id]: action.payload,
        },
      };

    case "user/handleAnswer":
      const updatedUserAnswer = Object.values(state.user).find(
        (user) => user.id === action.payload.id
      );

      const updatedQuestionAnswer = Object.values(state.listQuestion).find(
        (question) => question.id === action.payload.question
      );

      if (updatedUserAnswer && updatedUserAnswer.question) {
        updatedUserAnswer.question.push(action.payload.question);
        updatedUserAnswer.answers.push(action.payload.questionSelect);
      }

      if (action.payload.option === "optionOne") {
        updatedQuestionAnswer?.optionOne.votes.push(action.payload.id);
      } else {
        updatedQuestionAnswer?.optionTwo.votes.push(action.payload.id);
      }

      return {
        ...state,
        user: { ...state.user },
        listQuestion: { ...state.listQuestion },
      };

    default:
      return state;
  }
};

export default reducer;
