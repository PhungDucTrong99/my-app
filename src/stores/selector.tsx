export const todoListSelector = (state: any) => state.user;
export const listPolls = (state: any) => state.user;
export const listQuestion = (state: any) => state.listQuestion;
