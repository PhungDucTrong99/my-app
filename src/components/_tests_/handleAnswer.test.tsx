import { _saveQuestionAnswer } from "../../stores/action";
import { AnyAction } from "redux";

import configureMockStore, { MockStoreEnhanced } from "redux-mock-store";
import thunk, { ThunkDispatch } from "redux-thunk";

const middlewares = [thunk];
const mockStore = configureMockStore<
  unknown,
  ThunkDispatch<unknown, unknown, AnyAction>
>(middlewares);

interface User {
  id: string;
  username: string;
  image: any;
  password: string;
  role: string;
  status: string;
  totalPools: number;
  answers: string[];
  question: string[];
}

interface Question {
  id: string;
  author: string;
  createtime: string;
  optionOne: {
    votes: string[];
    text: string;
  };
  optionTwo: {
    votes: string[];
    text: string;
  };
}

export interface State {
  isLoading: {
    status: boolean;
  };
  user: { [key: string]: User };
  listQuestion: { [key: string]: Question };
}

export const initialState: State = {
  isLoading: {
    status: false,
  },
  user: {
    Cat: {
      id: "000",
      username: "Cat",
      image: "",
      password: "070699",
      role: "admin",
      status: "inActive",
      totalPools: 4,
      answers: ["become a super hero"],
      question: ["6ni6ok3ym7mf1p33lnez"],
    },
    Fox: {
      id: "111",
      username: "Fox",
      image: "",
      password: "070699",
      role: "admin",
      status: "inActive",
      totalPools: 1,
      answers: ["become a super hero"],
      question: ["6ni6ok3ym7mf1p33lnez"],
    },
    Gorilla: {
      id: "222",
      username: "Gorilla",
      role: "admin",
      image: "",
      password: "070699",
      status: "inActive",
      totalPools: 1,
      answers: ["become a super hero"],
      question: ["6ni6ok3ym7mf1p33lnez"],
    },
  },
  listQuestion: {
    "8xf0y6ziyjabvozdd253nd": {
      id: "8xf0y6ziyjabvozdd253nd",
      author: "000",
      createtime: "11:03 AM : 13/12/2013",
      optionOne: {
        votes: [""],
        text: "have horrible short term memory",
      },
      optionTwo: {
        votes: [""],
        text: "have horrible long term memory",
      },
    },
    "6ni6ok3ym7mf1p33lnez": {
      id: "6ni6ok3ym7mf1p33lnez",
      author: "111",
      createtime: "11:03 AM : 12/12/2023",
      optionOne: {
        votes: [""],
        text: "become a super hero",
      },
      optionTwo: {
        votes: [""],
        text: "become a supervillian",
      },
    },
    am8ehyc8byjqgar0jgpub9: {
      id: "am8ehyc8byjqgar0jgpub9",
      author: "222",
      createtime: "11:03 AM : 12/12/2022",
      optionOne: {
        votes: [""],
        text: "be telekinetic",
      },
      optionTwo: {
        votes: [""],
        text: "be telepathic",
      },
    },
    loxhs1bqm25b708cmbf3g: {
      id: "loxhs1bqm25b708cmbf3g",
      author: "111",
      createtime: "11:03 AM : 12/12/2011",
      optionOne: {
        votes: [""],
        text: "be a front-end developer",
      },
      optionTwo: {
        votes: [""],
        text: "be a back-end developer",
      },
    },
  },
};
describe("_saveQuestionAnswer", () => {
  it("should send the right action and update the status", async () => {
    const data = {
      id: "000",
      question: "8xf0y6ziyjabvozdd253nd",
      questionSelect: "become a super hero",
      option: "optionOne",
    };

    const initialState = {
      user: {
        Cat: {
          id: "000",
          username: "Cat",
          answers: ["become a super hero", "optionOne"],
          question: ["6ni6ok3ym7mf1p33lnez", "8xf0y6ziyjabvozdd253nd"],
        },
      },
      listQuestion: {
        "8xf0y6ziyjabvozdd253nd": {
          id: "8xf0y6ziyjabvozdd253nd",
          optionOne: {
            votes: ["000"],
            text: "become a super hero",
          },
          optionTwo: {
            votes: [""],
            text: "have horrible long term memory",
          },
        },
        "6ni6ok3ym7mf1p33lnez": {
          id: "6ni6ok3ym7mf1p33lnez",
          optionOne: {
            votes: [""],
            text: "become a super hero",
          },
          optionTwo: {
            votes: [""],
            text: "become a supervillian",
          },
        },
      },
    };

    const expectedActions = [
      {
        type: "user/handleAnswer",
        payload: data,
      },
    ];

    const store: MockStoreEnhanced<unknown, {}> = mockStore(initialState);

    await store.dispatch<any>(_saveQuestionAnswer(data)).then(() => {
      const actions = store.getActions();
      expect(actions).toEqual(expectedActions);

      const updatedState = store.getState() as typeof initialState;
      expect(updatedState.user["Cat"].question).toContain(
        "8xf0y6ziyjabvozdd253nd"
      );
      expect(updatedState.user["Cat"].answers).toContain("become a super hero");
      expect(
        updatedState.listQuestion["8xf0y6ziyjabvozdd253nd"].optionOne.votes
      ).toContain("000");
      expect(
        updatedState.listQuestion["8xf0y6ziyjabvozdd253nd"].optionOne.votes
      ).toMatchSnapshot();
    });
  });

  it("will return an error if data transmission is incorrect", async () => {
    const incorrectData = {
      id: "000",
      question: "invalidQuestionId",
      questionSelect: "become a super hero",
      option: "optionOne",
    };

    const store: MockStoreEnhanced<unknown, {}> = mockStore(initialState);
    let error: any = null;
    try {
      await store.dispatch<any>(_saveQuestionAnswer(incorrectData));

      expect(true).toBe(false);
    } catch (err: any) {
      error = err;
    }
    expect(error).toBeDefined();
    expect(error).toMatchSnapshot();
  });
});
