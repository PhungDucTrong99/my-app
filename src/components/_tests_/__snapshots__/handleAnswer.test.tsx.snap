// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_saveQuestionAnswer should dispatch the correct action and update the state 1`] = `
Array [
  "000",
]
`;

exports[`_saveQuestionAnswer should return an error if incorrect data is passed 1`] = `
[Error: [2mexpect([22m[31mreceived[39m[2m).[22mtoBe[2m([22m[32mexpected[39m[2m) // Object.is equality[22m

Expected: [32mfalse[39m
Received: [31mtrue[39m]
`;

exports[`_saveQuestionAnswer should send the right action and update the status 1`] = `
Array [
  "000",
]
`;

exports[`_saveQuestionAnswer will return an error if data transmission is incorrect 1`] = `
[Error: [2mexpect([22m[31mreceived[39m[2m).[22mtoBe[2m([22m[32mexpected[39m[2m) // Object.is equality[22m

Expected: [32mfalse[39m
Received: [31mtrue[39m]
`;
